lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+                cast<model::CallEdge>(Edge.get())->IsNoReturn = true;
+
+              Block.Successors.insert(Edge);
+            } else if (auto *Call = getCallTo(I, "function_call")) {
+              // Marker can be erased now, no longer needed.
+              Call->eraseFromParent();
+
+              auto Destination = getBasicBlockPC(Succ);
+              auto Edge = makeEdge(Destination,
+                                   model::FunctionEdgeType::FakeFunctionCall);
+              Block.Successors.insert(Edge);
+            } else {
+              Queue.insert(Succ);
+            }
+          }
+        }
+      }
+
+      Block.End = End + Size;
+      CFG.insert(Block);
+    }
+  }
+
+  return CFG;
+}
+
+template<class FO>
+FunctionSummary
+CFEPAnalyzer<FO>::analyze(GlobalsArrayRef ABIRegisters,
+                          llvm::BasicBlock *Entry,
+                          llvm::raw_fd_ostream &OutputIBI,
+                          llvm::raw_fd_ostream &OutputAAWriter) {
+  using namespace llvm;
+
+  Value *RA = nullptr;
+  const auto *PCH = GCBI->programCounterHandler();
+  StructType *MetaAddressTy = MetaAddress::getStruct(&M);
+
+  // Run function boundaries detection, and retrieve the outlined function.
+  Function *OutFunc = createDisposableFunction(Entry);
+
+  // Initialize CFG of the function for the model.
+  auto CFG = collectDirectCFG(Entry, OutFunc);
+
+  // Identify the callee-saved registers and tell if a function returns
+  // properly, i.e., it jumps to the return address. To achieve this, we craft
+  // the IR by loading the PC, the SP, as well as the ABI registers CSVs
+  // respectively at function entry and end. We win when the subtraction between
+  // their entry and end values is zero.
+  IRBuilder<> Builder(&OutFunc->getEntryBlock().front());
+  auto *IntPtrTy = GCBI->spReg()->getType();
+  auto *IntTy = GCBI->spReg()->getType()->getElementType();
+
+  auto *SP = Builder.CreateLoad(GCBI->spReg());
+  auto *SPPtr = Builder.CreateIntToPtr(SP, IntPtrTy);
+  auto *GEPI = Builder.CreateGEP(IntTy, SPPtr, ConstantInt::get(IntTy, 0));
+
+  switch (GCBI->arch()) {
+  case Triple::arm: {
+    constexpr auto *LinkRegister = "r14";
+    RA = Builder.CreateLoad(M.getGlobalVariable(LinkRegister, true));
+    break;
+  }
+
+  case Triple::aarch64: {
+    constexpr auto *LinkRegister = "lr";
+    RA = Builder.CreateLoad(M.getGlobalVariable(LinkRegister, true));
+    break;
+  };
+
+  case Triple::mips:
+  case Triple::mipsel: {
+    constexpr auto *ReturnRegister = "ra";
+    RA = Builder.CreateLoad(M.getGlobalVariable(ReturnRegister, true));
+    break;
+  }
+
+  case Triple::systemz: {
+    constexpr auto *ReturnRegister = "r14";
+    RA = Builder.CreateLoad(M.getGlobalVariable(ReturnRegister, true));
+    break;
+  }
+
+  case Triple::x86:
+  case Triple::x86_64: {
+    RA = Builder.CreateLoad(SPPtr);
+    break;
+  }
+
+  default:
+    revng_abort("Unsupported architecture");
+  }
```

Comment:
This makes me quite sad. This logic should be moved to `BinaryFile.cpp` and expose its result through the `Architecture` data structure.

At very least, outline this code and put a big `TODO`.

Reply:
Sorry that this makes you sad – yet, you might recollect that we discussed multiple times about moving this into `GCBI` or something, but yeah, still there.